### PR TITLE
feat: add minMetricSamplesToAlarm property

### LINKED ISSUE #198: Allows users specify minimal sample size below which the alarm will be suppressed and considered unreliable
### Feature scope

Alarming

### Describe your suggested feature

The need for this feature is most visible when alarming on the AVERAGE, P90, P99... of some metric, which might contain sparse datapoints and hence decrease the reliability of the aggregated value.

Example:
You have an alarm on the P99 Latency metric. But during nights the TPS is going very low, hence you don't get enough records to reliably understand if there is an issue increasing latency or it's because of regular cache misses, retries, etc... If during 1 minute you got only a few calls to the service, the P99 latency won't be reliable reprepsentation of the health of the system. Hence we don't want to alarm on it.

With the current setup of the monitoring constructs, this is not easily possible. We have to create the composite alarms manually, add the actions, deal with the sample size metrics, etc...

By adding this to the library, we'll allow users to create alarms that are closer to their needs.

Definition of done:
- By adding a field to the alarms definition, the users can define minimal sample size after which they want the alarm to be enabled
- The change should be backwards compatible

### PR DESCRIPTION
**Pull request by voho**
Fixes #198 

---

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license_

### CODE REVIEW

File: lib/common/alarm/AlarmFactory.ts [hunk 1]
```diff
@@ -496,20 +517,62 @@ export class AlarmFactory {
...
+
+    let alarm: AlarmBase = primaryAlarm;
+
+    // create composite alarm for min metric samples (if defined)
+
+    if (props.minMetricSamplesToAlarm) {
+      const metricSampleCount = metric.with({ statistic: MetricStatistic.N });
+      const noSamplesAlarm = metricSampleCount.createAlarm(
+        this.alarmScope,
+        alarmName + "-NoSamples",
```
Comment (echeung-amzn) [inline lib/common/alarm/AlarmFactory.ts:549]:
```suggestion
        `${alarmName}-NoSamples`,
```

File: lib/common/alarm/AlarmFactory.ts [hunk 2]
```diff
@@ -496,20 +517,62 @@ export class AlarmFactory {
...
+        {
+          alarmName: alarmName + "-NoSamples",
+          alarmDescription: "The metric does not have enough samples.",
```
Comment (echeung-amzn) [inline lib/common/alarm/AlarmFactory.ts:552]: Maybe something like "${Metric name} has insufficient samples."?

File: lib/common/alarm/AlarmFactory.ts (hunk 1 above)
Comment (echeung-amzn) [reply lib/common/alarm/AlarmFactory.ts:549]: Likewise for other string concatenations.

File: test/common/alarm/AlarmFactory.test.ts [hunk 3]
```diff
@@ -193,9 +193,9 @@ test("addAlarm: period override is propagated to alarm metric", () => {
     alarmNameSuffix: "TwoHoursPeriod",
     period: Duration.hours(2),
   });
-  const alarm1hConfig = alarm1h.alarm.metric.toMetricConfig();
+  const alarm1hConfig = (alarm1h.alarm as any).metric.toMetricConfig();
```
Comment (echeung-amzn) [inline test/common/alarm/AlarmFactory.test.ts:196]: Could cast as an `Alarm` for better type checking?

**Review by echeung-amzn — CHANGES_REQUESTED** [review]

File: lib/common/alarm/AlarmFactory.ts (hunk 1 above)
Comment (voho) [reply lib/common/alarm/AlarmFactory.ts:549]: Will fix

File: lib/common/alarm/AlarmFactory.ts (hunk 2 above)
Comment (voho) [reply lib/common/alarm/AlarmFactory.ts:552]: Yeah, maybe better to put more details there.

File: test/common/alarm/AlarmFactory.test.ts (hunk 3 above)
Comment (voho) [reply test/common/alarm/AlarmFactory.test.ts:196]: Good catch, will change.

**Review by echeung-amzn — APPROVED** [review]

**Pull request merged**